core/trino-spi/src/main/java/io/trino/spi/type/Timestamps.java

```diff
@@ -150,20 +157,29 @@ static String formatTimestampWithTimeZone(int precision, long epochMillis, int p
         long picoFraction = (long) floorMod(epochMillis, MILLISECONDS_PER_SECOND) * PICOSECONDS_PER_MILLISECOND + picosOfMilli;
         LocalDateTime dateTime = LocalDateTime.ofInstant(instant, zoneId);
 
-        return formatTimestamp(precision, dateTime, picoFraction)
-                .append(" ")
-                .append(zoneId.getId()).toString();
+        String zoneIdString = zoneId.getId();
+        StringBuilder builder = new StringBuilder(timestampFormatLength(precision) + zoneIdString.length() + 1);
+        formatTimestamp(precision, dateTime, picoFraction, builder);
+        return builder.append(' ').append(zoneIdString).toString();
     }
 
-    private static StringBuilder formatTimestamp(int precision, LocalDateTime dateTime, long picoFraction)
+    private static void formatTimestamp(int precision, LocalDateTime dateTime, long picoFraction, StringBuilder builder)
     {
-        StringBuilder builder = new StringBuilder();
-        builder.append(TIMESTAMP_FORMATTER.format(dateTime));
+        TIMESTAMP_FORMATTER.formatTo(dateTime, builder);
         if (precision > 0) {
             long scaledFraction = picoFraction / POWERS_OF_TEN[MAX_PRECISION - precision];
-            builder.append(".");
-            builder.append(format("%0" + precision + "d", scaledFraction));
+            builder.append('.');
+            builder.setLength(builder.length() + precision);
+            int index = builder.length() - 1;
+
+            // Append the fractional the decimal digits in reverse order
+            // comparable to format("%0" + precision + "d", scaledFraction);
+            for (int i = 0; i < precision; i++) {
+                long temp = scaledFraction / 10;
+                int digit = (int) (scaledFraction - (temp * 10));
```

Comment:
Why not `scaledFraction % 10` ? Is there a measurable performance difference?

Reply:
I didn't specifically measure it, this is a variation on the logic inside of `Long.toString` so I presume it was implemented that way on purpose.

(`Long.toString` resolves two decimal digits per iteration and therefore also uses lookup tables to convert to `char` instead of `(char)('0' + digit)`, but that was definitely overkill for this scenario).